src/slurm_cloud_bursting/suspend.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
+# the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import logging
+
+import argparse
+import boto3
+from botocore.config import Config
+from botocore.exceptions import ClientError
+from configparser import ConfigParser
+
+from slurm_cloud_bursting.utils import (
+    CONFIG_FILE_PATH,
+    DEFAULT_MAX_INSTANCES_BATCH_SIZE,
+    DEFAULT_MAX_RETRY,
+    SCONTROL,
+    get_nodes_info,
+    grouper,
+    run_command,
+)
+
+SlurmSuspendConfig = collections.namedtuple(
+    "SlurmSuspendConfig", ["region", "cluster_name", "boto3_config", "max_batch_size"],
+)
+SUSPEND_LOG_FILE = "/var/log/parallelcluster/slurm_suspend.log"
+
+
+def _get_config():
+    """
+    Get resume program configuration.
+
+    :return: the configuration parameters
+    """
+    logging.info("Reading %s", CONFIG_FILE_PATH)
+
+    config = ConfigParser()
+    config.read(CONFIG_FILE_PATH)
+    lvl = config.get("slurm_cb_config", "loglevel", fallback=logging.INFO)
+    logging.getLogger().setLevel(lvl)
+
+    boto3_config = {}
+    region = config.get("slurm_cb_config", "region")
+    cluster_name = config.get("slurm_cb_config", "cluster_name")
+    max_retry = int(config.get("slurm_cb_config", "max_retry", fallback=DEFAULT_MAX_RETRY))
+    if max_retry > 0:
+        boto3_config["retries"] = {
+            "max_attempts": max_retry,
+            "mode": "standard",
+        }
+    max_batch_size = int(config.get("slurm_cb_config", "max_batch_size", fallback=DEFAULT_MAX_INSTANCES_BATCH_SIZE))
+
+    proxy = config.get("slurm_cb_config", "proxy", fallback="NONE")
+    if proxy != "NONE":
+        boto3_config["proxies"] = {"https": proxy}
+
+    logging.info(
+        (
+            "Configured parameters: region=%s cluster_name=%s boto3_config=%s max_batch_size=%s",
+            region,
+            cluster_name,
+            boto3_config,
+            max_retry,
+            max_batch_size,
+        )
+    )
+    return SlurmSuspendConfig(region, cluster_name, Config(**boto3_config), max_batch_size)
+
+
+def delete_instances(instance_ids_to_nodename, region, boto3_config, batch_size):
+    """Terminate corresponding EC2 instances."""
+    ec2_client = boto3.client("ec2", region_name=region, config=boto3_config)
+    logging.info("Terminating instances %s", list(instance_ids_to_nodename.keys()))
+    for instances in grouper(instance_ids_to_nodename.keys(), batch_size):
+        try:
+            ec2_client.terminate_instances(InstanceIds=instances,)
```

Comment:
If you specify multiple instances and the request fails (for example, because of a single incorrect instance ID), none of the instances are terminated. Not sure if we could better handle the error conditions tbh. I'm wondering if we want to retry on failures at the very least (or verify on what error conditions boto3 client retries automatically)